Finish TODO's in ``contributing/typings.rst`` (#3545)
ManimCommunity/manim: docs/source/contributing/typings.rst

  pending shape support, using the correct type aliases will help users understand
  which shape should be used.

> * For typings for generic functions, check out `this <https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes>`_

Reviewer: ```suggestion
* For typings of generic collections, check out `this <https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes>`_
```

        return self.set_color(mobject.get_color())

* Always parametrize generics (``list[int]`` instead of ``list``,
  ``type[Any]`` instead of ``type``, etc.). This also applies to callables.

.. code:: py

    rate_func: Callable[[float], float] = lambda t: smooth(1 - t)

> * Use ``TypeVar`` when dealing with the same class but different instances.

Reviewer: ```suggestion
* Use ``TypeVar`` when you want to "link" type hints as being the same type.
```

Or similar

  Consider ``Mobject.copy``, which returns a new instance of the same class.
  It would be type-hinted as:

.. code:: py

>     T = TypeVar()

Reviewer: ```suggestion
    T = TypeVar("T")
```

    def copy(self: T) -> T:
        ...

* Use ``typing.Iterable`` whenever the function works with *any* iterable, not a specific type.

> * If the function returns a container of a specific length each time, use ``tuple``.

Reviewer: ```suggestion
* If the function returns a container of a specific length each time, consider using ``tuple`` instead of ``list``.
```

.. code:: py

   def foo() -> tuple[float, float, float]:
       return (0, 0, 0)

* If a function works with a parameter as long as said parameter has a ``__getitem__``, ``__iter___`` and ``__len__`` method,

>   the typehint of the parameter should be ``typing.Mapping``. If it also supports ``__setitem__`` and/or ``__delitem__``, it

Reviewer: ```suggestion
  the typehint of the parameter should be ``collections.abc.Mapping``. If it also supports ``__setitem__`` and/or ``__delitem__``, it
```

>   should be marked as ``typing.MutableMapping``.

Reviewer: ```suggestion
  should be marked as ``collections.abc.MutableMapping``.
```

> * Typehinting something as ``object`` means that variable should only access methods that are accessible to every python object,

Reviewer: ```suggestion
* Typehinting something as ``object`` means that only attributes available on every Python object should be accessed,
```

>   like ``__str__`` and so on. On the other hand, literally any method can be called on a variable with the ``Any`` typehint -

Reviewer: ```suggestion
  like ``__str__`` and so on. On the other hand, literally any attribute can be accessed on a variable with the ``Any`` typehint -
```

(methods are attributes)

  It would be type-hinted as:

.. code:: py

    T = TypeVar("T")

    def copy(self: T) -> T:
        ...

* Use ``typing.Iterable`` whenever the function works with *any* iterable, not a specific type.

* If the function returns a container of a specific length each time, consider using ``tuple`` instead of ``list``.

.. code:: py

   def foo() -> tuple[float, float, float]:
       return (0, 0, 0)

* If a function works with a parameter as long as said parameter has a ``__getitem__``, ``__iter___`` and ``__len__`` method,
  the typehint of the parameter should be ``collections.abc.Mapping``. If it also supports ``__setitem__`` and/or ``__delitem__``, it
  should be marked as ``collections.abc.MutableMapping``.

* Typehinting something as ``object`` means that only attributes available on every Python object should be accessed,
  like ``__str__`` and so on. On the other hand, literally any attribute can be accessed on a variable with the ``Any`` typehint -
  it's more freeing than the ``object`` typehint, and makes mypy stop typechecking the variable. Note that whenever possible,
  try to keep typehints as specific as possible.

> * If importing a module or a library purely for typehints, keep it under an ``if typing.TYPE_CHECKING`` guard, to prevent them from
>   being imported at runtime (helps library performance).

Reviewer: ```suggestion
* If objects are imported purely for type hint purposes, keep it under an ``if typing.TYPE_CHECKING`` guard, to prevent them from
  being imported at runtime (helps library performance). Do not forget to use the ``from __future__ import annotations`` import to avoid having runtime ``NameError`` exceptions.
```
